cmd/config/config.go

```diff
@@ -1,82 +1,178 @@
+	if cmd == "" {
+		usage(fs)
+		os.Exit(2)
+	}
+
+	if err := fs.Parse(args); err != nil {
+		fatal(err.Error())
+	}
+
+	if inplace {
+		out = in
+	} else if out == "[in].out" {
+		out = in + ".out"
+	}
+
+	key := []byte(keyStr)
+	var err error
+	switch cmd {
+	case "upgrade":
+		err = upgradeFile(in, out, key)
+	case "decrypt":
+		err = encryptWrapper(in, out, key, false, decryptFile)
+	case "encrypt":
+		err = encryptWrapper(in, out, key, true, encryptFile)
 	}
 
-	fileData, err := os.ReadFile(inFile)
 	if err != nil {
-		log.Fatalf("Unable to read input file %s. Error: %s.", inFile, err)
+		fatal(err.Error())
 	}
 
-	if config.ConfirmECS(fileData) && encrypt {
-		log.Println("File is already encrypted. Decrypting..")
-		encrypt = false
+	fmt.Println("Success! File written to " + out)
+}
+
+func upgradeFile(in, out string, key []byte) error {
+	c := &config.Config{
+		EncryptionKeyProvider: func(_ bool) ([]byte, error) {
+			if len(key) != 0 {
+				return key, nil
+			}
+			return config.PromptForConfigKey(false)
+		},
 	}
 
-	if !config.ConfirmECS(fileData) && !encrypt {
-		var result interface{}
-		errf := json.Unmarshal(fileData, &result)
-		if errf != nil {
-			log.Fatal(errf)
-		}
-		log.Println("File is already decrypted. Encrypting..")
-		encrypt = true
+	if err := c.ReadConfigFromFile(in, true); err != nil {
+		return err
 	}
 
-	var data []byte
-	if encrypt {
-		data, err = config.EncryptConfigFile(fileData, []byte(key))
-		if err != nil {
-			log.Fatalf("Unable to encrypt config data. Error: %s.", err)
-		}
-	} else {
-		data, err = config.DecryptConfigFile(fileData, []byte(key))
-		if err != nil {
-			log.Fatalf("Unable to decrypt config data. Error: %s.", err)
+	return c.SaveConfigToFile(out)
+}
+
+type encryptFunc func(string, []byte) ([]byte, error)
+
+func encryptWrapper(in, out string, key []byte, confirmKey bool, fn encryptFunc) error {
+	if len(key) == 0 {
+		var err error
+		if key, err = config.PromptForConfigKey(confirmKey); err != nil {
+			return err
 		}
 	}
+	outData, err := fn(in, key)
+	if err != nil {
+		return err
+	}
+	if err := file.Write(out, outData); err != nil {
+		return fmt.Errorf("unable to write output file %s; Error: %w", out, err)
+	}
+	return nil
+}
+
+func encryptFile(in string, key []byte) ([]byte, error) {
+	if config.IsFileEncrypted(in) {
+		return nil, errors.New("file is already encrypted")
+	}
+	outData, err := config.EncryptConfigFile(readFile(in), key)
+	if err != nil {
+		return nil, fmt.Errorf("unable to encrypt config data. Error: %w", err)
+	}
+	return outData, nil
+}
+
+func decryptFile(in string, key []byte) ([]byte, error) {
+	if !config.IsFileEncrypted(in) {
+		return nil, errors.New("file is already decrypted")
+	}
+	outData, err := config.DecryptConfigFile(readFile(in), key)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decrypt config data. Error: %w", err)
+	}
+	if json, err := jsonparser.Set(outData, []byte("-1"), "encryptConfig"); err == nil {
```

Comment:
```suggestion
	if json, err := jsonparser.Set(outData, []byte("-1"), "encryptConfig"); err != nil {
		// err probably means it didn't decrypt, but we don't tell the user that for security
		return nil, err
	} else {
		outData = json
	}
	return outData, nil
```

Will prevent the mangling file issue

Reply:
Just realised I added a comment to explain my reasoning for not returning the error☝️

Reply:
Fixed https://github.com/gbjk/gocryptotrader/commit/57dab4133